clang/lib/CodeGen/Targets/AArch64.cpp

```diff
@@ -307,7 +307,12 @@ AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadic,
     // 0.
     if (IsEmpty && Size == 0)
       return ABIArgInfo::getIgnore();
-    return ABIArgInfo::getDirect(llvm::Type::getInt8Ty(getVMContext()));
+    // An empty struct can have size greater than one byte if alignment is
+    // involved.
+    // When size <= 64, we still hold it by i8 in IR and lowering to registers.
+    // When Size > 64, just fall through to avoid va_list out of sync.
```

Comment:
"to avoid va_list out of sync" doesn't really make sense; the important thing is that this follows the ABI rule.

Maybe take another look at the suggested comment inhttps://github.com/llvm/llvm-project/pull/72197#issuecomment-1815284976 .

Reply:
Forgive my foolish，can I just copy it as comment?  

> AAPCS64 does not say that empty records are ignored as arguments,
> but other compilers do so in certain situations, and we copy that behavior.
> Those situations are in fact language-mode-specific, which seems really
> unfortunate, but it's something we just have to accept. If this doesn't
> apply, just fall through to the standard argument-handling path.
> Darwin overrides the psABI here to ignore all empty records in all modes.

Reply:
Yes, I intended that as a suggestion for the comment, and you're welcome to include it verbatim.

I agree with Eli that I'd like to understand why we need to have this `i8` carve-out.

Reply:
https://github.com/llvm/llvm-project/commit/9bb857a4f1aec1c33d599672defb46b98d792719#diff-d5392fbd34fbb607cc468738adce16985f9461c9a3f5bc7c5ed64dba44137d19R3752 This is I can found when  the `i8` was introduced. Guess @TNorthover know it.

Reply:
It looks like Tim is just implementing the standard rule from the PCS here and just didn't think about the fact that he could let the handling fall through to the main path.

`va_arg` would not require empty structs to occupy space for parameter passing; at worst, it would require us to recognize empty structs in the `va_arg` logic.  Even if it were unimplementable, though, that'd be a psABI problem, not a compiler one.